conn.go

```diff
@@ -32,18 +32,125 @@ import (
+func Open(opts ...ConnOption) (*Conn, error) {
+	cc := &Conn{}
+	for _, opt := range opts {
+		opt(cc)
+	}
+
+	cc.mu.Lock()
+	defer cc.mu.Unlock()
+	nlconn, err := cc.dialNetlink()
+	if err != nil {
+		return nil, err
+	}
+	cc.nlconn = nlconn
+	return cc, nil
+}
+
+// WithNetNSFd sets the network namespace to open a new netlink connection to:
+// the fd must reference a network namespace.
+func WithNetNSFd(fd int) ConnOption {
+	return func(cc *Conn) {
+		cc.NetNS = fd
+	}
+}
+
+// WithTestDial sets the specified nltest.Func when opening a new netlink
+// connection.
+func WithTestDial(f nltest.Func) ConnOption {
+	return func(cc *Conn) {
+		cc.TestDial = f
+	}
+}
+
+// netlinkCloser is returned by netlinkConn(UnderLock) and must be called after
+// being done with the returned netlink connection in order to properly close
+// this connection, if necessary.
+type netlinkCloser func() error
+
+// netlinkConn returns a netlink connection together with a netlinkCloser that
+// later must be called by the caller when it doesn't need the returned netlink
+// connection anymore. The netlinkCloser will close the netlink connection when
+// necessary. If Open has been called before then this "long-running" netlink
+// connection will be returned, otherwise a new "temporary" netlink connection
+// will be opened and returned instead. netlinkConn must not be called while the
+// Conn.mu lock is currently helt. Use netlinkConnUnderLock instead in such
+// situations.
+func (cc *Conn) netlinkConn() (*netlink.Conn, netlinkCloser, error) {
+	cc.mu.Lock()
+	defer cc.mu.Unlock()
+	return cc.netlinkConnUnderLock()
+}
+
+// netlinkConnUnderLock works like netlinkConn but must be called while holding
+// the Conn.mu lock.
+func (cc *Conn) netlinkConnUnderLock() (*netlink.Conn, netlinkCloser, error) {
+	if cc.nlconn != nil {
+		return cc.nlconn, func() error { return nil }, nil
+	}
+	nlconn, err := cc.dialNetlink()
+	if err != nil {
+		return nil, nil, err
+	}
+	return nlconn, func() error { return nlconn.Close() }, nil
+}
+
+// Close closes the (long-running) netlink connection that has been opened using
+// Open. If either no (long-running) netlink connection has been opened or this
+// connection is already in the process of closing, Close will immediately
+// return without any error.
+//
+// Please note that Close solely closes a long-running netlink connection, but
+// never closes any temporary netlink connection(s) that might be still alive at
+// the same time.
+func (cc *Conn) Close() error {
+	// Don't acquire the lock for the whole duration of the Close operation, but
+	// instead only to make sure to only run the netlink socket close on the
+	// first time with an open long-running netlink socket. As Open() acts as
+	// constructor, it is not possible to re-Open an already closed connection.
+	cc.mu.Lock()
+	nlconn := cc.nlconn
+	cc.nlconn = nil
+	cc.mu.Unlock()
+	if nlconn != nil {
+		var err error
+		cc.closeOnce.Do(func() {
```

Comment:
Isn’t closeOnce superfluous when you’re already using a mutex?

Reply:
yes, agreed ... but wasn't exactly sure to not overlook some corner case.

Reply:
removed superfluous closeOnce